rspace/src/main/scala/coop/rchain/rspace/internal.scala

```diff
@@ -84,4 +86,37 @@ object internal {
     (codecSeq(codecC) ::
       codecSeq(codecDatum(codecA)) ::
       codecSeq(codecWaitingContinuation(codecP, codecK))).as[GNAT[C, P, A, K]]
+
+  type MultisetMultiMap[K, V] = mutable.HashMap[K, Multiset[V]]
+
+  object MultisetMultiMap {
+    def empty[K, V]: MultisetMultiMap[K, V] = new mutable.HashMap[K, Multiset[V]]()
+  }
+
+  implicit class RichMultisetMultiMap[K, V](msmm: MultisetMultiMap[K, V]) {
```

Comment:
We could spare an allocation here by making this a value class

Reply:
Good call.